torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py

```diff
@@ -371,7 +371,18 @@ def aot_dispatch_autograd(
                 == len(fw_metadata.input_info) + inner_meta.num_outputs_rng_offset
             )
             for i, (bw_out) in enumerate(bw_outs):
-                if bw_out is None and _can_detach(_input_node(fx_g, i)):
+                # If our input experiences a metadata mutation inside the graph (e.g. set_()),
+                # we *must* not detach, otherwise it will be the detach'd input that gets the metadata mutation
+                metadata_mutation_in_graph = (
+                    fw_metadata.input_info[i].mutation_type
+                    == MutationType.MUTATED_IN_GRAPH
+                    and fw_metadata.input_info[i].mutates_storage_metadata
+                )
+                if (
+                    bw_out is None
+                    and not metadata_mutation_in_graph
```

Comment:
oh fwiw @jansel - this is probably what makes it ok for me to remove the `can_detach()` logic.

I guess it is morally similar to the `can_detach()` code, but it applies for any input metadata mutations that we keep in the graph: if there is a graph input that we mutate the metadata of (via `set_()`, or any other metadata mutation), we can't detach first, or else the user's tensor won't get the metadata mutation.